File: backend/app/api/v1/endpoints/creditors.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Optional
from uuid import UUID
from app.core.database import get_db
from app.schemas.creditor import CreditorCreate, CreditorUpdate, CreditorResponse
from app.services.creditor import CreditorService
from app.models.user import User
from app.api.dependencies.auth import get_current_user
from app.schemas.common import MessageResponse

router = APIRouter()

@router.get("/", response_model=List[CreditorResponse])
async def get_creditors(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    search: Optional[str] = Query(None),
    sort_by: Optional[str] = Query(None),
    sort_order: str = Query("asc"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    creditor_service = CreditorService(db)
    return await creditor_service.get_all(current_user.company_id, skip, limit, search, sort_by, sort_order)

@router.post("/", response_model=CreditorResponse)
async def create_creditor(
    data: CreditorCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    creditor_service = CreditorService(db)
    return await creditor_service.create(current_user.company_id, data, current_user.id)

@router.get("/{creditor_id}", response_model=CreditorResponse)
async def get_creditor(
    creditor_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    creditor_service = CreditorService(db)
    return await creditor_service.get_by_id(creditor_id, current_user.company_id)

@router.put("/{creditor_id}", response_model=CreditorResponse)
async def update_creditor(
    creditor_id: UUID,
    data: CreditorUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    creditor_service = CreditorService(db)
    return await creditor_service.update(creditor_id, current_user.company_id, data, current_user.id)

@router.delete("/{creditor_id}", response_model=MessageResponse)
async def delete_creditor(
    creditor_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    creditor_service = CreditorService(db)
    return await creditor_service.delete(creditor_id, current_user.company_id, current_user.id)

File: frontend/src/api/services/creditors.ts
import api from '../axios';

export interface Creditor {
  id: string;
  name: string;
  email: string | null;
  phone: string | null;
  address: string | null;
  tax_id: string | null;
  credit_limit: number;
  outstanding_balance: number;
  is_active: boolean;
  status: string;
}

export interface CreditorCreate {
  creditor_type: 'B2B' | 'B2C';
  creditor_code: string;
  business_type: string;
  name: string;
  email: string;
  phone: string;
  address?: string;
  gst_number?: string;
  credit_limit?: number;
}

export const creditorService = {
  getCreditors: async (params?: { search?: string; sort_by?: string; sort_order?: 'asc' | 'desc' }): Promise<Creditor[]> => {
    const response = await api.get('/creditors', { params });
    return response.data;
  },
  
  getCreditor: async (id: string): Promise<Creditor> => {
    const response = await api.get(`/creditors/${id}`);
    return response.data;
  },
  
  createCreditor: async (data: CreditorCreate): Promise<Creditor> => {
    const response = await api.post('/creditors', data);
    return response.data;
  },
  
  updateCreditor: async (id: string, data: Partial<CreditorCreate>): Promise<Creditor> => {
    const response = await api.put(`/creditors/${id}`, data);
    return response.data;
  },
  
  deleteCreditor: async (id: string): Promise<void> => {
    await api.delete(`/creditors/${id}`);
  }
};
